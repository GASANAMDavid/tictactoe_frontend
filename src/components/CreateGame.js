import Select from "react-select";
import Button from "./Button";
import { Form } from "react-bootstrap";
import { useState } from "react";

const languageOptions = [
  { value: "en", label: "English" },
  { value: "fr", label: "French" },
];

const boardSizeOptions = [
  { value: 3, label: 3 },
  { value: 4, label: 4 },
  { value: 5, label: 5 },
];

const CreateGame = ({
  onSelectLanguage,
  onCreateGame,
  translations,
  gameInfo,
  setGameInfo,
}) => {
  const [isLanguageSet, setIsLanguageSet] = useState(false);

  const modeOptions = translations['gameModes'].map((mode) => {
    return { value: mode, label: mode.label }
  })
  
  const handleChange = (lang) => {
    setGameInfo({ ...gameInfo, language: lang });
    onSelectLanguage(lang);
    setIsLanguageSet(true);
  };
  return (
    <div className='create-game'>
      {isLanguageSet ? (
        <Form>
          <Form.Row>
            <label>{translations.playerName}:</label>
            <input
              className='input-player-name'
              placeholder={translations.promptPlayerName}
              value={gameInfo.playerName}
              onChange={(event) => {
                setGameInfo({ ...gameInfo, playerName: event.target.value });
              }}
            />
          </Form.Row>
          <Form.Row>
            <label>{translations.symbol}:</label>
            <input
              className='input-symbol'
              value={gameInfo.symbol}
              placeholder={translations.promptSymbol}
              onChange={(event) =>
                setGameInfo({ ...gameInfo, symbol: event.target.value })
              }
            />
          </Form.Row>
          <Form.Row>
            <label>{translations.gameMode}:</label>
            <Select
              className='select-game-mode'
              placeholder={translations.select}
              options={modeOptions}
              value={modeOptions.find(option => option.value === gameInfo.gameMode)}
              onChange={(event) =>
                setGameInfo({ ...gameInfo, gameMode: event.value })
              }
            />
          </Form.Row>
          <Form.Row>
            <label>{translations.boardSize}:</label>
            <Select
              className='select-board-size'
              options={boardSizeOptions}
              placeholder={translations.select}
              onChange={(event) =>
                setGameInfo({ ...gameInfo, boardSize: event.value })
              }
              value={boardSizeOptions.find(option => option.value === gameInfo.boardSize)}
            />
          </Form.Row>
          <Button
            color='#2699ab'
            text={translations.newGame}
            onClick={() => onCreateGame()}
          />
        </Form>
      ) : (
        <label>
          Choose language
          <Select
            className='select-language'
            options={languageOptions}
            value={languageOptions.find(option => option.value === gameInfo.language)}
            onChange={(event) => {
              handleChange(event.value);
            }}
          />
        </label>
      )}
    </div>
  );
};

export default CreateGame;
