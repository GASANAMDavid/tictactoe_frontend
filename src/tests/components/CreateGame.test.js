// import React from "react";
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import CreateGame from "../../components/CreateGame";

describe("CreateGame", () => {
  let createGame;
  const mockTranslations = { gameModes: [] };
  const mockGetSelectedLanguageContent = jest.fn();
  const mockCreateGame = jest.fn();
  const mockSetGameInfo = jest.fn();
  const mockGameInfo = {};

  const lang = "en";

  beforeEach(() => {
    createGame = shallow(
      <CreateGame
        onSelectLanguage={mockGetSelectedLanguageContent}
        translations={mockTranslations}
        onCreateGame={mockCreateGame}
        gameInfo={mockGameInfo}
        setGameInfo={mockSetGameInfo}
      />
    );
    createGame.find(".select-language").simulate("change", { value: lang });
  });

  it("renders correctly", () => {
    expect(toJson(createGame)).toMatchSnapshot();
  });

  describe("when language is not set", () => {
    it("calls onSelectLanguage with the selected_language", () => {
      expect(mockGetSelectedLanguageContent).toHaveBeenCalledWith(lang);
    });
  });

  describe("when language is set", () => {
    it("should have input for player name", () => {
      createGame.find('.input-player-name').simulate('change', { target: { value: 'David' } })
      expect(mockSetGameInfo).toHaveBeenCalled()
      expect(createGame.find(".input-player-name").exists()).toBe(true);
    });

    it("should have input for symbol", () => {

      createGame.find('.input-symbol').simulate('change', { target: { value: 'David' } })
      expect(mockSetGameInfo).toHaveBeenCalled()
      expect(createGame.find(".input-symbol").exists()).toBe(true);
    });

    it("should allow user to choose game mode", () => {
      createGame.find('.select-game-mode').simulate('change', { target: { value: 'David' } })
      expect(mockSetGameInfo).toHaveBeenCalled()
      expect(createGame.find(".select-game-mode").exists()).toBe(true);
    });

    it("allows user to choose boardSize", () => {
      createGame.find('.select-board-size').simulate('change', { target: { value: 'David' } })
      expect(mockSetGameInfo).toHaveBeenCalled()
      expect(createGame.find(".select-board-size").exists()).toBe(true);
    });

    describe("submit button", () => {
      it("has a Button component to create game", () => {
        expect(createGame.find("Button").exists()).toBe(true);
      });

      it("call onCreateGame when button clicked", () => {
        createGame.find("Button").simulate("click");
        expect(mockCreateGame).toHaveBeenCalled();
      });
    });
  });
});
