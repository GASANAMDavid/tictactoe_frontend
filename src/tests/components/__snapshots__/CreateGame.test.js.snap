// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateGame renders correctly 1`] = `
<div
  className="create-game"
>
  <Form
    inline={false}
  >
    <FormRow>
      <label>
        :
      </label>
      <input
        className="input-player-name"
        onChange={[Function]}
      />
    </FormRow>
    <FormRow>
      <label>
        :
      </label>
      <input
        className="input-symbol"
        onChange={[Function]}
      />
    </FormRow>
    <FormRow>
      <label>
        :
      </label>
      <StateManager
        className="select-game-mode"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Human V Random Computer",
              "value": 1,
            },
            Object {
              "label": "Human V Intelligent Computer",
              "value": 2,
            },
          ]
        }
      />
    </FormRow>
    <FormRow>
      <label>
        :
      </label>
      <StateManager
        className="select-board-size"
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": 3,
              "value": 3,
            },
            Object {
              "label": 4,
              "value": 4,
            },
            Object {
              "label": 5,
              "value": 5,
            },
          ]
        }
      />
    </FormRow>
    <Button
      color="#2699ab"
      onClick={[Function]}
    />
  </Form>
</div>
`;
